models/tshirt/include/Tshirt.h

```diff
@@ -0,0 +1,289 @@
+        }
+
+        static double calc_Sfc(const tshirt_params& params, const tshirt_state& state)
+        {
+            // TODO: account for possiblity of Hwt being less than 0.5 (though initially, it looks like this will never be the case)
+            double z1 = calc_Hwt(params) - 0.5;
+            double z2 = z1 + 2;
+
+            // Note that z^( 1 - (1/b) ) / (1 - (1/b)) == b * (z^( (b - 1) / b ) / (b - 1)
+            return params.maxsmc * pow((1.0 / params.satpsi), (-1.0 / params.b)) *
+                   ((params.b * pow(z2, ((params.b - 1) / params.b)) / (params.b - 1)) -
+                    (params.b * pow(z1, ((params.b - 1) / params.b)) / (params.b - 1)));
+        }
+
+        static void init_nash_cascade_vector(vector<Nonlinear_Reservoir>& reservoirs, const tshirt_params& params,
+                const tshirt_state& state, const double activation, const double max_flow_velocity)
+        {
+            reservoirs.resize(params.nash_n);
+            for ( unsigned long i = 0; i < reservoirs.size(); ++i )
+            {
+                //construct a single outlet nonlinear reservoir
+                reservoirs[i] = Nonlinear_Reservoir(0, params.Ssmax, state.Snash[i], params.Kn, 1, activation,
+                        max_flow_velocity);
+            }
+        }
+
+        //! run one time step of tshirt
+        static int run(
+                double dt,
+                tshirt_params params,        //!< static parameters for tshirt
+                tshirt_state state,          //!< model state
+                tshirt_state& new_state,     //!< model state struct to hold new model state
+                tshirt_fluxes& fluxes,       //!< model flux object to hold calculated fluxes
+                double input_flux_meters,          //!< the amount water entering the system this time step
+                // TODO: should/can this be a smart pointer?
+                giuh_kernel* giuh_obj,       //!< kernel object for calculating GIUH runoff from subsurface lateral flow
+                void* et_params)            //!< parameters for the et function
+        {
+            double column_total_soil_moisture_deficit = params.Ssmax - state.Ss;
+
+            // Note this surface runoff value has not yet performed GIUH calculations
+            double surface_runoff, subsurface_infiltration_flux;
+
+            Schaake_partitioning_scheme(dt, calc_Cschaake(params), column_total_soil_moisture_deficit, input_flux_meters,
+                    &surface_runoff, &subsurface_infiltration_flux);
+
+            double Sfc = calc_Sfc(params, state);
+
+            vector<Reservoir_Outlet> subsurface_outlets;
+
+            // Keep track of the indexes of the specific outlets for later access
+            int lf_outlet_index = 0;
+            int perc_outlet_index = 1;
+
+            // TODO: confirm these are correct
+            // Max transmissivity
+            double max_lateral_flow = params.satdk * params.multiplier * params.Ssmax;
```

Comment:
again, anything derived strictly from params should be relatively static, can be defined at initilization and resued by the instance of the model class

Reply:
I have done this with this and a few other such derived values.